### PR TITLE
Improve runtime compatibility, capture extraction, and token-sequence execution

### PR DESCRIPTION
**Pull request by jbachorik**
# Summary

This PR improves Reggie's runtime compatibility and performance for large capture-heavy regexes, especially Grok-style patterns that expand into long regular expressions with many named groups.

The work spans parser compatibility, allocation-friendly capture extraction, capture projection, safer fallback behavior, DFA scalability improvements, and a new structural execution path for deterministic token-sequence patterns.

# Major changes

## Parser/API compatibility

- Supports atomic groups `(?>...)` by parsing them as non-capturing groups. For Reggie's non-backtracking engines, atomic-group backtracking-prevention semantics are equivalent to `(?:...)`.
- Supports `\Q...\E` quoted literals, including inside character classes.
- Adds public `UnsupportedPatternException` for callers that need precise unsupported-pattern handling.

## Allocation-friendly capture extraction

- Adds `matchInto(String input, int[] groupStarts, int[] groupEnds)` for caller-owned capture-boundary arrays.
- Adds corresponding find/match-into support across runtime and generated paths.
- Preserves the contract that caller arrays remain unchanged on no-match.

## Named capture projection

- Adds `CapturePolicy` and `ReggieOptions`.
- Adds `CapturePolicy.NAMED_ONLY` to preserve original named-group indexes while dropping unobservable unnamed captures from runtime capture tracking.
- Keeps named-group lookup compatible with callers that discover original group indexes from the source pattern and later read by index.

## DFA/codegen scalability and fallback behavior

- Adds DFA state-budget handling for oversized state spaces.
- Adds a table-driven DFA backend.
- Improves generated-method-too-large diagnostics with class/method/descriptor/code-size details.
- Fixes lookaround/fallback behavior needed by large expanded patterns.

## Structural token-sequence execution path

Adds a generic structural route:

```text
regex AST -> PatternCategorizer -> LinearTokenSequencePlan -> LinearTokenSequenceMatcher
```

This route recognizes reusable deterministic token atoms such as:

- literals and whitespace
- non-space fields
- IP/host-like fields
- quoted fields
- delimiter captures
- signed integers and decimals
- optional token subsequences
- trailing bounded bracketed-word captures

The route is structural: it does not depend on exact pattern strings or specific capture names.

## Correctness and regression coverage

- Adds parser compatibility tests.
- Adds capture-policy and match-into API tests.
- Adds structural categorizer and linear-token-sequence planner/runtime tests.
- Adds real expanded Grok-style regex fixtures as regression tests.
- Adds JDK/Reggie named-capture boundary equivalence tests for representative large capture-heavy patterns, including optional fields and delimiter/quoted-field edge cases.
- Adds/extends routing/fallback tests and algorithmic fuzzing infrastructure.

# Performance validation

Representative integrated Grok-style benchmark with native token-sequence routing:

| Engine | Score | Allocation |
|---|---:|---:|
| JDK regex | 16.210 ± 2.128 us/op | 7701.393 ± 154.805 B/op |
| Reggie native token sequence | 2.353 ± 0.161 us/op | 7682.979 ± 61.845 B/op |

Coverage for the benchmarked patterns:

```text
2/2 native, 0/2 internal JDK fallback, 0/2 supplier JDK fallback
```

# Validation

```bash
./gradlew spotlessApply build
```

Passed.

### CODE REVIEW

File: reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/LinearTokenSequenceMatcher.java [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+              ends,
+              i == op.children().size() - 1,
+              optionalDepth + 1);
+      if (next < 0) {
+        System.arraycopy(savedStarts, 0, starts, 0, starts.length);
+        System.arraycopy(savedEnds, 0, ends, 0, ends.length);
+        return pos;
+      }
+    }
+    return next;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/LinearTokenSequenceMatcher.java:381]: Addressed by rejecting ambiguous optional token-sequence plans such as `(?:a|)a` from the fast path, so they fall back to the general regex route instead of committing to the present optional branch. Added regression coverage in `LinearTokenSequenceMatcherTest`.

File: reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/RuntimeCompiler.java [hunk 2]
```diff
@@ -139,11 +161,23 @@ public static Set<String> cachedPatterns() {
...
     try {
       // 1. Parse pattern to AST
       RegexParser parser = new RegexParser();
       RegexNode ast = parser.parse(pattern);
       Map<String, Integer> nameMap = parser.getGroupNameMap();
+      if (options.capturePolicy() == CapturePolicy.NAMED_ONLY) {
+        ReggieMatcher linearTokenSequenceMatcher =
+            tryCompileLinearTokenSequence(pattern, ast, nameMap);
+        if (linearTokenSequenceMatcher != null) {
+          return linearTokenSequenceMatcher;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/RuntimeCompiler.java:178]: Addressed by applying `CaptureProjection.preserveNamedAndSemanticCaptures(...)` before attempting linear token-sequence routing under `CapturePolicy.NAMED_ONLY`. Added a regression test verifying unnamed groups remain unset on the token-sequence path.

File: reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/LinearTokenSequenceMatcher.java [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+    int digitStart = pos;
+    while (pos < input.length() && isDigit(input.charAt(pos))) pos++;
+    if (pos == digitStart) return -1;
+    set(starts, ends, group, start, pos);
+    return pos;
+  }
+
+  private static int captureSignedIntegerOrDash(
+      String input, int pos, int group, int[] starts, int[] ends) {
+    if (pos < input.length() && input.charAt(pos) == '-') return pos + 1;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-runtime/src/main/java/com/datadoghq/reggie/runtime/LinearTokenSequenceMatcher.java]: Addressed by distinguishing captured dash alternations from alternations where only the integer branch is captured. Captured `(?:-|[+-]?\\d+)` now records `-`, while `(?>(?<bytes>[+-]?\\d+)|-)` still leaves the group unset for the dash branch to match JDK semantics. Added regression coverage.

File: AGENTS.md [hunk 4]
```diff
@@ -425,6 +425,28 @@ Verify both:
...
+- `PatternInfo` subclass field added → add it to that class's `structuralHashCode()`
+
+Example — `acceptanceAnchorConditions` and `entryGuard` added post-anchor fix:
+```java
+// DFAState: per-state acceptance anchor conditions
+hash = 31 * hash + state.acceptanceAnchorConditions.hashCode();
+
+// DFATransition: per-transition entry guard
+hash = 31 * hash + entry.getValue().entryGuard.hashCode();
+```
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply AGENTS.md:457]: Addressed in `a0873c1`: updated the Structural Hash Rule example to use an ordinal-derived `anchorBitmask(...)` for anchor EnumSets instead of `EnumSet.hashCode()`.

File: reggie-runtime/src/test/java/com/datadoghq/reggie/runtime/AnchorDiagTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+          (jdkFound == reggieFound)
+              && (!jdkFound || (jm.start() == r.start() && jm.end() == r.end()));
+      System.out.printf(
+          "%s  pat=%-20s inp=%-5s jdk=%s reggie=%s class=%s%n",
+          ok ? "OK  " : "FAIL",
+          pat,
+          "\"" + inp + "\"",
+          jdkFound ? "[" + jm.start() + "," + jm.end() + ")" : "null",
+          reggieFound ? "[" + r.start() + "," + r.end() + ")" : "null",
+          rm.getClass().getSimpleName());
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-runtime/src/test/java/com/datadoghq/reggie/runtime/AnchorDiagTest.java]: Addressed in `a0873c1`: converted `AnchorDiagTest` from stdout-only diagnostics into real JUnit assertions against JDK `find()` result and span.

File: reggie-runtime/src/test/java/com/datadoghq/reggie/runtime/AnchorDiagTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        (jdkFound == reggieFound)
+            && (!jdkFound || (jm.start() == r.start() && jm.end() == r.end()));
+    System.out.printf(
+        "%s  pat=%-25s inp=%-8s jdk=%-12s reggie=%-12s strategy=%s%n",
+        ok ? "OK  " : "FAIL",
+        pat,
+        "\"" + inp + "\"",
+        jdkSpan,
+        reggieSpan,
+        rm.getClass().getSimpleName());
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-runtime/src/test/java/com/datadoghq/reggie/runtime/AnchorDiagTest.java]: Addressed in `a0873c1`: `check()` now delegates to an assertion helper, so Reggie/JDK divergence fails the test instead of only printing `FAIL`.
